remote-add: Add --sign-verify=KEYTYPE=[inline|file]:PUBKEYREF

cgwalters opened the pull request:
Per https://github.com/ostreedev/ostree/issues/2080#issuecomment-623614483

A huge benefit of ed25519 (and ECC in general) is that keys are very
short - short enough that it's completely reasonable to inline
them into a command line argument.

And I think that's a good model; it makes the keys very visible.

For example, someone could easily copy-paste a commandline
argument from a webpage (secured via TLS) that says to run
`ostree remote add --verify-ed25519=KEY`.

jlebon commented on src/ostree/ot-remote-builtin-add.c:
```diff
@@ -46,6 +47,7 @@ static GOptionEntry option_entries[] = {
   { "set", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_set, "Set config option KEY=VALUE for remote", "KEY=VALUE" },
   { "no-gpg-verify", 0, 0, G_OPTION_ARG_NONE, &opt_no_gpg_verify, "Disable GPG verification", NULL },
   { "no-sign-verify", 0, 0, G_OPTION_ARG_NONE, &opt_no_sign_verify, "Disable signature verification", NULL },
+  { "verify-ed25519", 0, 0, G_OPTION_ARG_STRING, &opt_verify_ed25519, "Verify signatures ed25519 public key PUBKEY", "PUBKEY" },
```
```suggestion
  { "verify-ed25519", 0, 0, G_OPTION_ARG_STRING, &opt_verify_ed25519, "Verify signatures using ed25519 public key string", "PUBKEY" },
```

?

cgwalters replied:
I wish github offered "squash suggestion into commit, add Co-authored-by"

d4s commented:
@cgwalters agree, it is a good idea to add remote with a key.
However I don't think `--verify-ed25519` option is a good name for usage.
It is too `ed25519`-centric, while the signapi for now is able to operate with several signing engines.
Please check options in [src/ostree/ot-builtin-sign.c](https://github.com/ostreedev/ostree/blob/master/src/ostree/ot-builtin-sign.c) as instance -- if we use option `--sign-type` for engine selection, we may pass-through the key, file with keys or basedir as additional option. And in this case it works for **any** engine we have, so do not need to provide engine-specific things in `ostree` builtins.

cgwalters commented:
> However I don't think --verify-ed25519 option is a good name for usage. It is too ed25519-centric, while the signapi for now is able to operate with several signing engines.

Yes, I split this out of a more generic patch to support `--signapi-verify=TYPE=KEY`...hm, OK I'll do that instead.

cgwalters commented:
> Yes, I split this out of a more generic patch to support --signapi-verify=TYPE=KEY...hm, OK I'll do that instead.

Oh right I remember why I switched now - we don't know that new key types will be ergonomic to add to the commandline inline.  So it'd have to be more like:

`remote add --sign-verify=<keytype>=inline:<key>`
`remote add --sign-verify=<keytype>=file:/path/to/file`
or something?

cgwalters commented:
OK reworked this - @d4s what do you think?

d4s commented on tests/test-signed-pull.sh:
```diff
@@ -98,13 +98,30 @@ ${CMD_PREFIX} ostree --repo=repo config set 'remote "origin"'.verification-dummy
+    assert_not_reached "remote add with invalid keytype succeeded"
+fi
+assert_file_has_content err.txt 'Requested signature type is not implemented'
+if ${CMD_PREFIX} ostree --repo=repo remote add other --sign-verify=ed25519 http://localhost 2>err.txt; then
```
1. Please move all `ed25519` related tests below -- after `has_sign_ed25519` check -- in other case it throws an error if there is no ed25519 support: `error: Requested signature type is not implemented`
2. And as far as you there -- please add the test for `dummy` engine with inlined key -- it should to work everywhere.

cgwalters replied:
OK, done.
